### ISSUE TITLE
Looking for a more convenient way to getter value of the upper/lower triangular matrix into 1D

### ISSUE THREAD

**Issue by Kiyosora** [pytorch/pytorch#49227]
## 🚀 Feature
Looking for a new function like `torch.triu_values` / `torch.tril_values` to gatter the value of the upper/lower triangular matrix into 1D shape more convenient.

## Motivation

When we need to gatter the value of an upper/lower triangular matrix into 1D shape, in NumPy way, we can easily use `triu_indices`/`tril_indices` to do it, such:
```
>>> a=np.arange(0,9).reshape(3,3)
>>> a
array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])
>>> a[np.triu_indices(3)]
array([0, 1, 2, 4, 5, 8])
```
But in PyTorch, It's seem that we can’t get the results by the previous way:
```
>>> t=torch.arange(0,9).reshape([3,3])
>>> t
tensor([[0, 1, 2],
        [3, 4, 5],
        [6, 7, 8]])
>>> t[torch.triu_indices(3,3)]
tensor([[[0, 1, 2],
         [0, 1, 2],
         [0, 1, 2],
         [3, 4, 5],
         [3, 4, 5],
         [6, 7, 8]],
        [[0, 1, 2],
         [3, 4, 5],
         [6, 7, 8],
         [3, 4, 5],
         [6, 7, 8],
         [6, 7, 8]]])
```
While I can think of other ways to achieve the purpose, but it's not convenient enough.
```
>>> t[torch.triu_indices(3,3)[0], torch.triu_indices(3,3)[1]]
tensor([0, 1, 2, 4, 5, 8])
```

## Pitch

I was wondering if there was an easier way to do it, or if we could implement a new function like `torch.triu_values` / `torch.tril_values` to get the value of the upper/lower triangular matrix more convenient, Just like:
```
>>> t=torch.arange(0,9).reshape([3,3])
>>> t
tensor([[0, 1, 2],
        [3, 4, 5],
        [6, 7, 8]])
>>> torch.triu_values(t)
tensor([0, 1, 2, 4, 5, 8])
>>> torch.tril_values(t)
tensor([0, 3, 4, 6, 7, 8])
```

## Alternatives
I'd like to work on it if this feature sounds feasible 😄

**Comment by SsnL** [comment]
`x[torch.tril_indices(3,3).unbind()]`

But yeah, we should think about numpy parity.

**Comment by mruberry** [comment]
Thank you for reporting this issue, @Kiyosora. It would be nice to be compatible with NumPy. I added some labels to clarify this will need design to verify we don't break any currently relied-upon behavior.

**Comment by franchesoni** [comment]
This doesn't work for a batch of matrices! :cry:

**Comment by ezyang** [comment]
cc @zdevito

**Comment by codgas** [comment]
@franchesoni 
For me just creating a mask worked for batches,  you can do like this 
`   mask = torch.ones(x.shape[3], Y.shape[3])
    x = x[:,mask.triu()==1]  
`

**Comment by charliebudd** [comment]
I would be in favour of this being in the API. Here's my implementation for batched matrices for those who arrive here looking for a snippet…
``` Python
def tril_flatten(tril):
    N = tril.size(-1)
    indicies = torch.tril_indices(N, N)
    indicies = N * indicies[0] + indicies[1]
    return tril.flatten(-2)[..., indicies]
```
You only need to swap out the indices function for the upper triangle variant…
``` Python
def _tri_flatten(tri, indicies_func):
    N = tri.size(-1)
    indicies = indicies_func(N, N)
    indicies = N * indicies[0] + indicies[1]
    return tri.flatten(-2)[..., indicies]

def tril_flatten(tril):
    return _tri_flatten(tril, torch.tril_indices)

def triu_flatten(triu):
    return _tri_flatten(triu, torch.triu_indices)
```

**Comment by Dreamupers** [comment]
x[torch.tril_indices(3,3).split(1)]
x[torch.triu_indices(3,3).split(1)]
